Random GSP compilation failure in Grails application when including Micronaut

### Expected Behavior

We have random build failures in our CI caused by an AST transformation included by Micronaut. Typically, running the build again with no changes will fix the issue.

The failure will occur on randomly different GSP's that don't really seem to share anything in common. I'm assuming it must be some kind of race condition.

I wouldn't expect Micronaut to care at all about GSP compiliation.

### Actual Behaviour

The GSP is different, but the stack is exactly the same:
```
23:59:46.810 [pool-1-thread-1] ERROR org.grails.gsp.compiler.GroovyPageCompiler - Error Compiling GSP File: _rejected.gsp - BUG! exception in phase 'semantic analysis' in source unit 'gsp_common_redelivery_emails_rejected_gsp.groovy' null
Exception in thread "main" java.util.concurrent.ExecutionException: BUG! exception in phase 'semantic analysis' in source unit 'gsp_common_redelivery_emails_rejected_gsp.groovy' null
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at org.grails.gsp.compiler.GroovyPageCompiler.compile(GroovyPageCompiler.groovy:138)
	at org.grails.web.pages.GroovyPageCompilerForkTask.compile(GroovyPageCompilerForkTask.groovy:102)
	at org.grails.web.pages.GroovyPageCompilerForkTask.run(GroovyPageCompilerForkTask.groovy:155)
	at org.grails.web.pages.GroovyPageCompilerForkTask.main(GroovyPageCompilerForkTask.groovy:106)
Caused by: BUG! exception in phase 'semantic analysis' in source unit 'gsp_common_redelivery_emails_rejected_gsp.groovy' null
	at org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:905)
	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:610)
	at org.grails.gsp.compiler.GroovyPageCompiler$_compileGSP_closure4.doCall(GroovyPageCompiler.groovy:234)
	at jdk.internal.reflect.GeneratedMethodAccessor54.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:428)
	at org.codehaus.groovy.runtime.IOGroovyMethods.withStream(IOGroovyMethods.java:1184)
	at org.codehaus.groovy.runtime.ResourceGroovyMethods.withInputStream(ResourceGroovyMethods.java:1882)
	at org.grails.gsp.compiler.GroovyPageCompiler.compileGSP(GroovyPageCompiler.groovy:208)
	at jdk.internal.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:101)
	at org.grails.gsp.compiler.GroovyPageCompiler$_compile_closure1.doCall(GroovyPageCompiler.groovy:120)
	at jdk.internal.reflect.GeneratedMethodAccessor17.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at groovy.lang.Closure.call(Closure.java:406)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at io.micronaut.ast.groovy.TypeElementVisitorTransform.visit(TypeElementVisitorTransform.groovy:85)
	at org.codehaus.groovy.transform.ASTTransformationVisitor.lambda$addPhaseOperationsForGlobalTransforms$4(ASTTransformationVisitor.java:337)
	at org.codehaus.groovy.control.CompilationUnit$ISourceUnitOperation.doPhaseOperation(CompilationUnit.java:901)
	... 42 more
```

The NPE occurs here:
https://github.com/micronaut-projects/micronaut-core/blob/079ec811a0e6637445535dad534e622b49e7ed61/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorTransform.groovy#L85

Since loaded visitors is static and being checked for null on line 80, it seems like it might be getting set to null in between the check and the usage, which creates a very small race condition.

Perhaps some kind of synchronization should be used here since it has static state and compilation can occur in parallel in some cases.

### Steps To Reproduce

You'll have to take this one on faith, as I can't find any way to reproduce it. It seems like some kind of random race condition with the groovy compiler. The best option might be to defensively code around it. We have been unable to find any predictable way to reproduce this problem.

It probably happens to us about 1 out of every 15 builds.

### Environment Information

Ubuntu 20.04.3 LTS
Java 11.0.13-zulu
Grails 5.0.1 (although this also happened with Grails 4.0.12 as well)
Micronaut 3.0.3

### Example Application

_No response_

### Version

3.0.3

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorEnd.groovy
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorEnd.groovy
@@ -53,7 +53,7 @@ class TypeElementVisitorEnd implements ASTTransformation, CompilationUnitAware {
 
     @Override
     void visit(ASTNode[] nodes, SourceUnit source) {
-        Map<String, LoadedVisitor> loadedVisitors = TypeElementVisitorTransform.loadedVisitors
+        Map<String, LoadedVisitor> loadedVisitors = TypeElementVisitorTransform.loadedVisitors.get()
 
         ClassWriterOutputVisitor classWriterOutputVisitor = null
         if (source.classLoader instanceof InMemoryByteCodeGroovyClassLoader) {
@@ -78,7 +78,7 @@ class TypeElementVisitorEnd implements ASTTransformation, CompilationUnitAware {
             }
         }
 
-        final List<AbstractBeanDefinitionBuilder> beanDefinitionBuilders = TypeElementVisitorTransform.beanDefinitionBuilders
+        final List<AbstractBeanDefinitionBuilder> beanDefinitionBuilders = TypeElementVisitorTransform.beanDefinitionBuilders.get()
         if (beanDefinitionBuilders) {
             File classesDir = compilationUnit.configuration.targetDirectory
             if (classWriterOutputVisitor == null) {
@@ -102,8 +102,8 @@ class TypeElementVisitorEnd implements ASTTransformation, CompilationUnitAware {
             classWriterOutputVisitor.finish()
         }
 
-        TypeElementVisitorTransform.loadedVisitors = null
-        TypeElementVisitorTransform.beanDefinitionBuilders.clear()
+        TypeElementVisitorTransform.loadedVisitors.remove()
+        TypeElementVisitorTransform.beanDefinitionBuilders.remove()
         AstAnnotationUtils.invalidateCache()
         AbstractAnnotationMetadataBuilder.clearMutated()
     }

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorStart.groovy
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorStart.groovy
@@ -52,7 +52,7 @@ class TypeElementVisitorStart implements ASTTransformation, CompilationUnitAware
 
     @Override
     void visit(ASTNode[] nodes, SourceUnit source) {
-        Map<String, LoadedVisitor> loadedVisitors = TypeElementVisitorTransform.loadedVisitors
+        Map<String, LoadedVisitor> loadedVisitors = TypeElementVisitorTransform.loadedVisitors.get()
 
         if (loadedVisitors == null) {
             loadedVisitors = [:]
@@ -120,7 +120,7 @@ class TypeElementVisitorStart implements ASTTransformation, CompilationUnitAware
             }
         }
 
-        TypeElementVisitorTransform.loadedVisitors = loadedVisitors
+        TypeElementVisitorTransform.loadedVisitors.set(loadedVisitors)
     }
 
     @Override

--- a/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorTransform.groovy
+++ b/inject-groovy/src/main/groovy/io/micronaut/ast/groovy/TypeElementVisitorTransform.groovy
@@ -68,21 +68,21 @@ import static org.codehaus.groovy.ast.ClassHelper.makeCached
 class TypeElementVisitorTransform implements ASTTransformation, CompilationUnitAware {
 
     private static ClassNode generatedNode = new ClassNode(Generated)
-    protected static Map<String, LoadedVisitor> loadedVisitors = null
-    protected static List<AbstractBeanDefinitionBuilder> beanDefinitionBuilders = []
+    protected static ThreadLocal<Map<String, LoadedVisitor>> loadedVisitors = new ThreadLocal<>()
+    protected static ThreadLocal<List<AbstractBeanDefinitionBuilder>> beanDefinitionBuilders = ThreadLocal.withInitial({ -> [] })
     private CompilationUnit compilationUnit
 
     @Override
     void visit(ASTNode[] nodes, SourceUnit source) {
         ModuleNode moduleNode = source.getAST()
         List<ClassNode> classes = moduleNode.getClasses()
-
-        if (loadedVisitors == null) return
+        Map<String, LoadedVisitor> visitors = loadedVisitors.get()
+        if (visitors == null) return
 
         GroovyVisitorContext visitorContext = new GroovyVisitorContext(source, compilationUnit)
         for (ClassNode classNode in classes) {
             if (!(classNode instanceof InnerClassNode && !Modifier.isStatic(classNode.getModifiers())) && classNode.getAnnotations(generatedNode).empty) {
-                Collection<LoadedVisitor> matchedVisitors = loadedVisitors.values().findAll { v ->
+                Collection<LoadedVisitor> matchedVisitors = visitors.values().findAll { v ->
                     v.matches(classNode)
                 }
 
@@ -105,7 +105,7 @@ class TypeElementVisitorTransform implements ASTTransformation, CompilationUnitA
             }
         }
 
-        beanDefinitionBuilders.addAll(visitorContext.getBeanElementBuilders())
+        beanDefinitionBuilders.get().addAll(visitorContext.getBeanElementBuilders())
     }
 
     @Override

